source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <string>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+/**
+ * This is a specialized structure intended for static header maps, but
+ * there may be other use cases.
+ * The structure is:
+ * 1. a length-based lookup table so only keys the same length as the
+ * target key are considered.
+ * 2. a trie that branches on the "most divisions" position of the key.
+ *
+ * For example, if we consider the case where the set of headers is
+ * `x-prefix-banana`
+ * `x-prefix-babana`
+ * `x-prefix-apple`
+ * `x-prefix-pineapple`
+ * `x-prefix-barana`
+ * `x-prefix-banaka`
+ *
+ * A standard front-first trie looking for `x-prefix-banana` would walk
+ * 7 nodes through the tree, first for `x`, then for `-`, etc.
+ *
+ * This structure first jumps to matching length, eliminating in this
+ * example case apple and pineapple.
+ * Then the "best split" node is on
+ *   `x-prefix-banana`
+ *               ^
+ * so the first node has 3 non-miss branches, n, b and r for that position.
+ * Down that n branch, the "best split" is on
+ *   `x-prefix-banana`
+ *                 ^
+ * which has two branches, n or k.
+ * Down the n branch is the leaf node (only `x-prefix-banana` remains) - at
+ * this point a regular string-compare checks if the key is an exact match
+ * for the string node.
+ */
+template <class Value> class CompiledStringMap {
+  using FindFn = std::function<Value(const absl::string_view&)>;
+
+public:
+  using KV = std::pair<std::string, Value>;
+  /**
+   * Returns the value with a matching key, or the default value
+   * (typically nullptr) if the key was not present.
+   * @param key the key to look up.
+   */
+  Value find(const absl::string_view& key) const {
+    if (key.size() >= table_.size() || table_[key.size()] == nullptr) {
+      return {};
+    }
+    return table_[key.size()](key);
+  };
+  /**
+   * Construct the lookup table. This is a somewhat slow multi-pass
+   * operation - using this structure is not recommended unless the
+   * table is initialize-once, use-many.
+   * @param initial a vector of key->value pairs.
+   */
+  void compile(std::vector<KV> initial) {
+    if (initial.empty()) {
+      return;
+    }
+    size_t longest = 0;
+    for (const KV& pair : initial) {
+      longest = std::max(pair.first.size(), longest);
+    }
+    table_.resize(longest + 1);
+    std::sort(initial.begin(), initial.end(),
+              [](const KV& a, const KV& b) { return a.first.size() < b.first.size(); });
+    auto it = initial.begin();
+    for (size_t i = 0; i <= longest; i++) {
+      auto start = it;
+      while (it != initial.end() && it->first.size() == i) {
```

Comment:
I am guessing we don't consider compiling to be important to be fast. But this nested loop seems like it might be slower than needed.

How about this for an algo:
```
flat_hash_map<int, std::vector<std::reference<Value>> size_to_values_map;
for (const KV& pair : initial) {
  size_to_values_map[pair.second.size()] = pair.second;
}
```
I realize this creates a data structure which doesn't map the ones you are using, but it seems like a good way to go get the elements organized by size. After that I'm not sure how this would fit into your algo.

I'm having a little trouble following the algorithm  also. What causes `it` to re-examine the entries that don't match size 'i'?

Maybe some more comment would help, or breaking down your compile flow into a few helper methods that might be more self-documenting.

Reply:
It was already not *really* a nested loop, in that the two indexes were moving "in parallel" (i.e. it wasn't n*m, it was n+m), but I've changed it to make that a bit clearer. Though still not *much* clearer since the std::find_if still looks like it might be an m-length lookup. But it's not!